marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -215,93 +225,81 @@ instance ToJSON Utxo where
+data UtxoResult = UtxoResult_
+  { utxoResultAddress          :: !C.AddressAny
+  , utxoResultTxIn             :: !C.TxIn
+  , utxoResultDatum            :: !(Maybe C.ScriptData) -- datumhash_datum
+  , utxoResultDatumHash        :: !(Maybe (C.Hash C.ScriptData))
+  , utxoResultValue            :: !C.Value
+  , utxoResultInlineScript     :: !(Maybe C.ScriptInAnyLang)
+  , utxoResultInlineScriptHash :: !(Maybe C.ScriptHash)
+  , utxoResultTxIndexInBlock   :: !TxIndexInBlock
+  --
+  , utxoResultSlotNo           :: !C.SlotNo
+  , utxoResultBlockHeaderHash  :: !(C.Hash C.BlockHeader)
+  -- bla
+  , utxoResultSpentInfo        :: !(Maybe SpentInfo)
+  } deriving (Eq, Show)
+
+instance Ord UtxoResult where
+  compare u u' = compare (utxoResultAddress u, utxoResultTxIn u) (utxoResultAddress u', utxoResultTxIn u')
 
-urCreationSlotNo :: Lens' UtxoRow C.SlotNo
-urCreationSlotNo = urCreationPoint . cpSlotNo
-
-urCreationBlockHash :: Lens' UtxoRow (C.Hash C.BlockHeader)
-urCreationBlockHash = urCreationPoint . cpBlockHash
+urSpentSlotNo :: Traversal' UtxoRow C.SlotNo
+urSpentSlotNo = urSpentInfo . _Just . siSlotNo
 
 urSpentBlockHash :: Traversal' UtxoRow (C.Hash C.BlockHeader)
-urSpentBlockHash = urSpentInfo . _Just . siSpentPoint . cpBlockHash
+urSpentBlockHash = urSpentInfo . _Just . siBlockHeaderHash
 
 urSpentTxId :: Traversal' UtxoRow C.TxId
 urSpentTxId = urSpentInfo . _Just . siSpentTxId
 
 instance FromJSON UtxoRow where
     parseJSON (Object v) = let
-
         parseSpentInfo = do
             s <- v .:? "spentSlotNo"
             bh <- v .:? "spentBlockHeaderHash"
             tId <- v .:? "spentTxId"
-            pure $ case (s, bh, tId) of
-                (Nothing, Nothing, Nothing)     -> Nothing
-                (Just s', Just bh', Just txId') -> Just $ SpentInfo (ChainPointRow s' bh') txId'
-                _error                          -> fail "Inconsistent spent info"
+            pure $ if
+              | Just s' <- s, Just bh' <- bh, Just txId' <- tId -> Just $ SpentInfo s' bh' txId'
+              | otherwise                                       -> Nothing
 
         in UtxoRow
             <$> v .: "utxo"
-            <*> (ChainPointRow <$> v .: "slotNo" <*> v .: "blockHeaderHash")
+            <*> v .: "slotNo"
+            <*> v .: "blockHeaderHash"
             <*> parseSpentInfo
     parseJSON _ = mempty
 
 instance ToJSON UtxoRow where
   toJSON ur = object
     [ "utxo" .= view urUtxo ur
-    , "slotNo" .= view urCreationSlotNo ur
-    , "blockHeaderHash" .= view urCreationBlockHash ur
+    , "slotNo" .= _urSlotNo ur
+    , "blockHeaderHash" .= _urBlockHeaderHash ur
     , "spentSlotNo" .= preview urSpentSlotNo ur
     , "spentBlockHeaderHash" .= preview urSpentBlockHash ur
     , "spentTxId" .= preview urSpentTxId ur
     ]
 
 data instance StorableResult UtxoHandle
-    = UtxoResult { getUtxoResult :: ![UtxoRow] }
+    = UtxoResult { getUtxoResult :: [UtxoResult] }
```

Comment:
Use strict? `![UtxoResult]`

Reply:
fb05023dc (the rest of the list will still be lazy though)